lib/fusion_jwt_authentication/jwks_strategy.ex

```diff
@@ -0,0 +1,127 @@
+defmodule FusionJWTAuthentication.JWKS_Strategy do
+  @moduledoc """
+  Contain strategy to fetch the JWKS if TokenJWKS is choosen as token_verifier.
+  Automatically fetch the jwks when the application is started.
+  Refetch the jwks once each time a kid isn't found inside the jwks.
+  The jwks is cached in an ets table
+  """
+
+  require Logger
+
+  alias __MODULE__.EtsCache
+  alias Joken.Signer
+  alias FusionJWTAuthentication.API.JWT
+
+  use GenServer, restart: :permanent
+
+  def start_link(arg) do
+    GenServer.start_link(__MODULE__, arg, name: __MODULE__)
+  end
+
+  @impl true
+  def init(_) do
+    {:ok, %{}, {:continue, :do_init}}
+  end
+
+  @impl true
+  def handle_continue(:do_init, _state) do
+    EtsCache.new()
+    fetch_and_cache_signers()
+    {:noreply, %{}}
+  end
+
+  @impl true
+  def handle_call(:fetch_signers, _from, state) do
+    fetch_and_cache_signers()
+    {:reply, nil, state}
+  end
+
+  def match_signer_for_kid(kid) do
+    with [{^kid, signer}] <- EtsCache.get_signer(kid) do
+      {:ok, signer}
+    else
+      [] -> fetch_signer_and_match_kid(kid)
+      err -> err
+    end
+  end
+
+  def fetch_signer_and_match_kid(kid) do
+    GenServer.call(__MODULE__, :fetch_signers, 10000)
+
+    with [{^kid, signer}] <- EtsCache.get_signer(kid) do
+      {:ok, signer}
+    else
+      _ -> {:error, :kid_does_not_match}
+    end
+  end
+
+  @doc "Fetch signers and save them in the ets table"
+  def fetch_and_cache_signers() do
+    with {:ok, {_status_code, %{"keys" => keys}}} <- JWT.jwks(),
+         {:ok, signers} <- validate_and_parse_keys(keys) do
+      Logger.debug("Fetched signers. #{inspect(signers)}")
+      EtsCache.put_signers(signers)
+    else
+      {:error, _reason} = err ->
+        Logger.warn("#{__MODULE__} failed to fetch signers. Reason: #{inspect(err)}")
+
+      err ->
+        Logger.warn("#{__MODULE__} got an unexpected error while fetching signers. Reason: #{inspect(err)}")
+    end
+  end
+
+  defp validate_and_parse_keys(keys) when is_list(keys) do
+    Enum.reduce_while(keys, {:ok, %{}}, fn key, {:ok, acc} ->
```

Comment:
Probably best todo {:ok, []} here and then later on concat keys to the acc like `[{key["kid"], signer} | acc]`.

And then with put_signers you can just insert the entire list into ets at once instead of looping and inserting one at a time.

Should be a lot better performance not using Map.put and only having to call :ets.insert instead of looping over all the items.